### ISSUE TITLE
Help on ideas for gracefully shutting down tasks spawned by set_delegate

### ISSUE THREAD

**Issue by 2snEM6** [amqp-rs/lapin#374]
I am using Lapin with Tokio, and I am making use of the `set_delegate` method to consume and execute messages in parallel. That works great, the problem arises when I want to shutdown the client gracefully, waiting for the tasks to finish.

Some ideas I had:

- Keep track of each delegate execution join handles and join when shutting down.
- Keep track of the number of running tasks with an (atomic?) counter, and wait for it to be 0, with a timeout to ignore stalled tasks.

Has anyone encountered a similar situation and perhaps could shed some light on how to accomplish this? 

Thank you in advance.

I will post a code snipped of my current setup as soon as a I have my laptop with me.

**Comment by Keruspe** [comment]
Maybe use a Barrier?

**Comment by 2snEM6** [comment]
> Maybe use a Barrier?

What I did in the end is not using `set_delegate` and use a while loop + spawn tasks using a [JoinSet](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.shutdown):

```rust
let tasks = JoinSet::new();

// On each received delivery, spawn a new task and add it to the JoinSet
while let Some(delivery_result) = cons.next().await {
    tasks.spawn(async {
        // Do something async
    })
}

// Something triggers a shutdown (Typically a SIGTERM signal)

// Wait until all tasks are done
while tasks.join_next().await.is_some() {}

// Shutdown
```

This works great, however both while loops prevent outer code from advancing. I wonder if there is a way of iterating through the received messages while doing something else. I.e:

```rust
// Some code
queue.listen().await // Iterates through received messages and spawns tasks (as seen above)

// Instead of blocking above, here do something else
println!("Do something else")
```
